python/ql/lib/semmle/python/frameworks/Flask.qll

```diff
@@ -408,21 +408,21 @@ module Flask {
     }
   }
 
+  private API::Node requestFileStorage() {
+    // TODO: This approach for identifying member-access is very adhoc, and we should
+    // be able to do something more structured for providing modeling of the members
+    // of a container-object.
+    result = request().getMember("files").getASubscript()
+    or
+    result = request().getMember("files").getMember("get").getReturn()
+    or
+    result = request().getMember("files").getMember("getlist").getReturn().getASubscript()
+  }
+
   /** An `FileStorage` instance that originates from a flask request. */
   private class FlaskRequestFileStorageInstances extends Werkzeug::FileStorage::InstanceSource {
     FlaskRequestFileStorageInstances() {
-      // TODO: This approach for identifying member-access is very adhoc, and we should
-      // be able to do something more structured for providing modeling of the members
-      // of a container-object.
-      exists(API::Node files | files = request().getMember("files") |
-        this.asCfgNode().(SubscriptNode).getObject() =
-          files.getAValueReachableFromSource().asCfgNode()
-        or
-        this = files.getMember("get").getACall()
-        or
-        this.asCfgNode().(SubscriptNode).getObject() =
-          files.getMember("getlist").getReturn().getAValueReachableFromSource().asCfgNode()
-      )
+      this = requestFileStorage().getAValueReachableFromSource()
```

Comment:
This seems wrong. For something like

```py
x = request.files['foo']
func(x)
```

we only want ONE `FlaskRequestFileStorageInstances`, but using `getAValueReachableFromSource` would make _both_ the RHS of the assignment, and the use of `x` a `FlaskRequestFileStorageInstances` right?

Reply:
The equality means that they _are_ both instances. However, seeing as this is extending `InstanceSource` and will likely be the starting point of type tracking, it is better to use `asSource` here.

Reply:
exactly my point :blush: